crates/qasm2/src/lex.rs

```diff
@@ -0,0 +1,772 @@
+    fn new(
+        source: Box<dyn std::io::BufRead + Send>,
+        filename: std::ffi::OsString,
+        strict: bool,
+    ) -> Self {
+        TokenStream {
+            filename,
+            strict,
+            source,
+            line_buffer: Vec::with_capacity(80),
+            done: false,
+            // The first line is numbered "1", and the first column is "0".  The counts are
+            // initialised like this so the first call to `next_byte` can easily detect that it
+            // needs to extract the next line.
+            line: 0,
+            col: 0,
+            try_version: false,
+            peeked: None,
+        }
+    }
+
+    /// Create a [TokenStream] from a string containing the OpenQASM 2 program.
+    pub fn from_string(string: String, strict: bool) -> Self {
+        TokenStream::new(
+            Box::new(std::io::Cursor::new(string)),
+            "<input>".into(),
+            strict,
+        )
+    }
+
+    /// Create a [TokenStream] from a path containing the OpenQASM 2 program.
+    pub fn from_path<P: AsRef<Path>>(path: P, strict: bool) -> Result<Self, std::io::Error> {
+        let file = std::fs::File::open(path.as_ref())?;
+        Ok(TokenStream::new(
+            Box::new(std::io::BufReader::new(file)),
+            Path::file_name(path.as_ref()).unwrap().into(),
+            strict,
+        ))
+    }
+
+    /// Read the next line into the managed buffer in the struct, updating the tracking information
+    /// of the position, and the `done` state of the iterator.
+    fn advance_line(&mut self) -> PyResult<usize> {
+        if self.done {
+            Ok(0)
+        } else {
+            self.line += 1;
+            self.col = 0;
+            self.line_buffer.clear();
+            // We can assume that nobody's running this on ancient Mac software that uses only '\r'
+            // as its linebreak character.
+            match self.source.read_until(b'\n', &mut self.line_buffer) {
+                Ok(count) => {
+                    if count == 0 || self.line_buffer[count - 1] != b'\n' {
+                        self.done = true;
+                    }
+                    Ok(count)
+                }
+                Err(err) => {
+                    self.done = true;
+                    Err(QASM2ParseError::new_err(message_generic(
+                        Some(&Position::new(&self.filename, self.line, self.col)),
+                        &format!("lexer failed to read stream: {}", err),
+                    )))
+                }
+            }
+        }
+    }
+
+    /// Get the next character in the stream.  This updates the line and column information for the
+    /// current bute as well.
+    fn next_byte(&mut self) -> PyResult<Option<u8>> {
+        if self.col >= self.line_buffer.len() && self.advance_line()? == 0 {
+            return Ok(None);
+        }
+        let out = self.line_buffer[self.col];
```

Comment:
Done in 3c0db6116.  This also made me realise that the handling I had for the version numbering would fail if there were comments between the `OPENQASM` token and the version identifier (although that's pathological formatting), so I improved the handling of that in d5d7115a5.